### PR TITLE
feat: add confirm password field and show passwords feature to change password popUp

### LINKED ISSUE #76: Add confirm password field and see password feature to change password popUp.
## Description
As a user I would like to see what password I'm typing and would like to confirm it, so that i don't have to change password again incase of a typo while changing the password.

### Update [Required]
- [ ] Added confirm password feature
- [ ] Added see password feature
 

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

## Estimation
1 hour

### PR DESCRIPTION
**Pull request by bharat-1809**
### Description
Add confirm password field with validation and toggle password visibility button on the change password popUp

Fixes #76 

### Flutter Channel:
- [ ] I have used the Flutter Beta channel on my local machine

### Type of Change:

- Code

**Code/Quality Assurance Only**

- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Physical Device
Steps:
- Go to the Settings page by clicking on the icon at the top right corner of the homepage.
- Click on the Change Password button to see whether the feature works properly or not
flutter_01
 

### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented on my code or provided relevant documentation, particularly in hard-to-understand areas

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Comment by techno-disaster** [comment]
@bharat-1809 can you attach a video showing how the new change password dialog works now?

**Comment by bharat-1809** [comment]
> @bharat-1809 can you attach a video showing how the new change password dialog works now?

Sure

**Comment by bharat-1809** [comment]
[Link to the video](https://drive.google.com/file/d/16QvpD4gXotQCotlEhw51dfLQ9eWWn880/view?usp=sharing)

**Comment by techno-disaster** [comment]
> [Link to the video](https://drive.google.com/file/d/16QvpD4gXotQCotlEhw51dfLQ9eWWn880/view?usp=sharing)

Please try to attach a gif next time onwards :)

**Comment by bharat-1809** [comment]
> > [Link to the video](https://drive.google.com/file/d/16QvpD4gXotQCotlEhw51dfLQ9eWWn880/view?usp=sharing)
> 
> Please try to attach a gif next time onwards :)

Sure😅

**Review by bartekpacia — APPROVED** [review]
looks good, thank you🎉

**Comment by techno-disaster** [comment]
The confirm password and show password features work as expected, but due to a prior bad merge, the change password dialog box dosen't respond correctly and the snackbar isn't showed. (BuildContext error). #83 fixes this. So lets merge this PR after that to confirm if the whole change password function works. Thanks for your work @bharat-1809

**Comment by bharat-1809** [comment]
Okay cool, thanks😃

File: lib/screens/settings/settings_screen.dart
Comment (bartekpacia) [inline lib/screens/settings/settings_screen.dart]: please use double quotes (`"`)

**Review by bartekpacia — CHANGES_REQUESTED** [review]
Nice!

File: lib/screens/settings/settings_screen.dart
Comment (bharat-1809) [reply lib/screens/settings/settings_screen.dart]: Okay sure

**Review by bartekpacia — DISMISSED** [review]

**Review by techno-disaster — CHANGES_REQUESTED** [review]
@bharat-1809 we follow a theme in our app, where if you click to show one of the show passsword eye icons all the fields show the password, can you do the same in the change password dialog window fields? Thank you

**Comment by bharat-1809** [comment]
Yeah sure, I'll definitely do it. I have a little doubt, whether the eye icon will be displayed on all the fields or only one on the topmost field?
Thanks :)

**Comment by techno-disaster** [comment]
> Yeah sure, I'll definitely do it. I have a little doubt, whether the eye icon will be displayed on all the fields or only one on the topmost field?
> Thanks :)

on all fields :)

**Comment by bharat-1809** [comment]
@Techno-Disaster I have made the required changes please have a look.
Thanks

Here's a gif of the same:

**Review by bartekpacia — DISMISSED** [review]

**Comment by bharat-1809** [comment]
@bartekpacia Is that review dismissal automatic? because I didn't do it. :sweat_smile:

**Comment by bartekpacia** [comment]
@bharat-1809 I don't know 👀
Maybe it's GitHub that does it automatically when the PR owner pushes new changes to the already-approved PRs.

**Comment by bharat-1809** [comment]
> @bharat-1809 I don't know 
> Maybe it's GitHub that does it automatically when the PR owner pushes new changes to the already-approved PRs.

Oh..okay :sweat_smile:

**Review by robotjellyzone — DISMISSED** [review]
I have tested & reviwed this PR and LGTM 👍 

Screenshot_1592645352

**Comment by techno-disaster** [comment]
Added status on hold because of a prior bad merge mentioned [here](https://github.com/anitab-org/mentorship-flutter/pull/92#issuecomment-640323243)

**Comment by techno-disaster** [comment]
@bharat-1809 could you resolve this conflict? would be happy to merge it then :) You might have to take changes from both the head and your change-pass branch. Feel free to ask for help here, incase you have any questions

**Comment by bharat-1809** [comment]
> @bharat-1809 could you resolve this conflict? would be happy to merge it then :) You might have to take changes from both the head and your change-pass branch. Feel free to ask for help here, incase you have any questions

Sure :)

**Comment by bharat-1809** [comment]
**@anitab-org/mentorship-flutter/maintainers** I have resolved the conflict, please have a look :)
Thanks

File: lib/screens/settings/settings_screen.dart
Comment (techno-disaster) [inline lib/screens/settings/settings_screen.dart:122]: @bharat-1809 did you test the apk? iirc topContext is needed or the app crashes when you click on submit

**Review by techno-disaster — CHANGES_REQUESTED** [review]

File: lib/screens/settings/settings_screen.dart
Comment (bharat-1809) [reply lib/screens/settings/settings_screen.dart:122]: @Techno-Disaster sorry my bad. I have now checked and resolved the issue and its working fine. Please have a look :)
Thanks

**Review by techno-disaster — APPROVED** [review]
LGTM good work :)

**Comment by techno-disaster** [comment]
@anitab-org/qa-team @anitab-org/mentorship-flutter-maintainers Can someone review this?

**Review by HaripriyaB — DISMISSED** [review]

**Review by robotjellyzone — DISMISSED** [review]
I have already tested this PR here --> https://github.com/anitab-org/mentorship-flutter/pull/92#pullrequestreview-434416174

so, LGTM 🎉  thanks for your contribution @bharat-1809 👍 

ALSO, please check/mark that "**I have used the Flutter Beta channel on my local machine**" section in the description of this PR .

**Review by techno-disaster — CHANGES_REQUESTED** [review]
Fix merge conflict

**Comment by bharat-1809** [comment]
> Fix merge conflict

Its done. @Techno-Disaster please have a look :)
Thanks

**Comment by techno-disaster** [comment]
Can someone from @anitab-org/qa-team give this a test? it was tested before but to confirm nothing broke after the merging the conflict can someone give this another check? Thank You

**Review by robotjellyzone — CHANGES_REQUESTED** [review]
The changes made in this PR were tested locally. Following are the results:

Code review - Done By @Techno-Disaster 

All possible responses (positive and negative tests) were tested as below:

Tested confirm password field , its validation with show passwords feature to change password popUp:

Screenshot/gif:
 1. Allow updating the password even when **new password is not matching with confirm password** : 

tested-pr

2. **Not Showing missing confirm password field message even if its empty** :

password-missing


Expected Result : 
-  When we write a password in confirm password field then it should match with the new password field.
-  if left empty, it should reflect it as a missing field and should not allow to update the password until it gets all the field values in the required form.
 
Actual Result : 

-  It is not correctly validating as it still allows the change in password even when both new & confirm password fields are not matching. 
-  Also, not showing any missing field error/popup even when confirm password field is empty

Additional testcases covered: N/A

Additional Comments: @bharat-1809 You need to make changes in your code and see , where the validation condition is breaking as both of the failed tests were only due to breaking validation case.

Status of PR  :  **Changes Requested**
Android Version: 10.0, Device: Android Emulator Pixel 3 XL API 29

**Comment by bharat-1809** [comment]
@anitab-org/maintainers Please have a look. The validation errors have been fixed

**Review by HaripriyaB — APPROVED** [review]

**Review by robotjellyzone — APPROVED** [review]
Now its working as --->

1. Showing missing password field --> 

Screenshot_1594744902

2. Also, shows appropriate validation message when passwords don't match -->

Screenshot_1594744935

Great job 👍 @bharat-1809

**Comment by isabelcosta** [comment]
@Techno-Disaster @bartekpacia as code owners can you review this please?

**Review by techno-disaster — APPROVED** [review]
LGTM

**Review by isabelcosta — APPROVED** [review]
Thank you for your contribution @bharat-1809 👏 🎉

**Pull request merged**
